### PR TITLE
Add types for protocols

### PR DESCRIPTION
**Pull request by ffflorian**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/protocols/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/**
+ * @param input The input url.
+ * @param first If `true`, the first protocol will be returned. If number, it will represent the zero-based index of the protocols array.
+ * @returns The array of protocols or the specified protocol.
+ */
+declare function protocols(input: string, first?: boolean | number): string | string[];
+declare function protocols(input: string, first: true | number): string;
+declare function protocols(input: string, first?: false): string[];
```
Comment (sheetalkamat) [inline types/protocols/index.d.ts]: These should be 
```ts
declare function protocols(input: string, first?: false): string[];
declare function protocols(input: string, first: true | number): string;
```

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/protocols/index.d.ts (hunk 1 above)
Comment (ffflorian) [reply types/protocols/index.d.ts]: Thanks, fixed!

File: types/protocols/protocols-tests.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import protocols = require('protocols');
+
+protocols('git+ssh://git@some-host.com/and-the-path/name'); // $ExpectType string | string[]
```
Comment (sheetalkamat) [inline types/protocols/protocols-tests.ts]: This should have `$ExpectType string[]`

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/protocols/protocols-tests.ts (hunk 2 above)
Comment (ffflorian) [reply types/protocols/protocols-tests.ts]: Fixed, thanks!

File: types/protocols/protocols-tests.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+import protocols = require('protocols');
+
+protocols('git+ssh://git@some-host.com/and-the-path/name'); // $ExpectType string
```
Comment (sheetalkamat) [inline types/protocols/protocols-tests.ts]: This is incorrect. Expected type needs to be string[]

**Review by sheetalkamat — CHANGES_REQUESTED** [review]

File: types/protocols/protocols-tests.ts (hunk 3 above)
Comment (ffflorian) [reply types/protocols/protocols-tests.ts]: Sorry :smiley: Should be good now.

**Pull request merged**